Allow RevertSam to output by read group

dshiga opened the pull request:
### Summary of changes

RevertSam's default behavior is unchanged, but you can now optionally output by read group. When run with OUTPUT_BY_READGROUP=true, multiple output bams will be created, each containing the reads for a single read group.

### Usage details

When outputting by read group, the user must provide either OUTPUT -- which must be a directory where the output files will be written (with default names of the form [read_group_id].bam) -- or OUTPUT_MAP, which is a two-column tab-separated file providing the output paths for each read group id.

### Implementation

When outputting by read group, we make a collection of writers -- one per read group -- and send each read to the appropriate writer. When not outputting by read group we send all reads to a single writer, as always.

When sorting and outputting by read group, we use a collection of sorters. When not outputting by read group, we use a single sorter, as always.

nh13 commented:
👎 

It's fewer than [50 lines of code in Python](https://gist.github.com/seandavi/2014542) to split into one file per read group, and then you can run `RevertSam` on each, what am I missing?

yfarjoun commented:
@nh13 
while I don't know the details behind the need for this functionality (@ktibbett ?) I can think of a few comments to your remark:

1. The fact that it "can be done" outside of picard isn't reason enough to exclude it from picard (IMHO). 
2. This solution solves in a single pass what yours does in two, so in that regard it's better.

nh13 commented:
1. The purpose of the python script example was to show that it can be done much more simply, and that the current PR is too complex (how many new classes?).
2. That requirement was never mentioned anywhere.  With all the sorting collections it is effectively buffering to disk with large input, so is it really a single pass?

Nonetheless, I think this one needs clear motivation and careful review as you can see my first instinct is to question the need for so much new code.

nh13 commented on src/java/picard/sam/revertsam/OutputMap.java:
```diff
@@ -0,0 +1,82 @@
+package picard.sam.revertsam;
```
license for all non-test sources.

nh13 commented on src/java/picard/sam/revertsam/OutputMap.java:
```diff
@@ -0,0 +1,82 @@
+ * Represents the mapping from each read group id to the corresponding output file
+ * that RevertSam is expected to write to.
+ */
+public class OutputMap {
```
Why is this class public by the constructors package protected?  Shouldn't the class be package protected?

nh13 commented on src/java/picard/sam/revertsam/OutputMap.java:
```diff
@@ -0,0 +1,82 @@
+public class OutputMap {
+    final Map<String, File> idToOutput = new HashMap<String, File>();
+
+    OutputMap(final File outputMapFile) {
```
Why not simplify and have two methods that return a `Map<String,File>` that stores for each read group id the corresponding paths?

nh13 commented on src/java/picard/sam/revertsam/OutputMap.java:
```diff
@@ -0,0 +1,82 @@
+        return new OutputMapEntry(id, outputPath);
+    }
+    
+    private static void validateRow(final String id, final String output, final File outputMapFile) {
```
this is not used elsewhere, why not inline it?

nh13 commented on src/java/picard/sam/revertsam/OutputMap.java:
```diff
@@ -0,0 +1,82 @@
+        return true;
+    }
+
+    private static OutputMapEntry createEntry(final TabbedTextFileWithHeaderParser.Row row, final File outputMapFile) {
```
why is this a separate function, inline it?

nh13 commented on src/java/picard/sam/revertsam/OutputMap.java:
```diff
@@ -0,0 +1,82 @@
+        final String id = row.getField("READ_GROUP_ID");
+        final String output = row.getField("OUTPUT");
+        validateRow(id, output, outputMapFile);
+        File outputPath = new File(output);
```
`final`?

nh13 commented on src/java/picard/sam/revertsam/OutputMap.java:
```diff
@@ -0,0 +1,82 @@
+        }
+    }
+
+    private static class OutputMapEntry {
```
Use a `Map<String,File>`?

nh13 commented on src/java/picard/sam/revertsam/PerReadGroupSorter.java:
```diff
@@ -0,0 +1,52 @@
+package picard.sam.revertsam;
```
I'm not complaining about the missing license.

nh13 commented on src/java/picard/sam/revertsam/PerReadGroupSorter.java:
```diff
@@ -0,0 +1,52 @@
+        sorter.add(rec);
+    }
+
+    private SortingCollection<SAMRecord> getSorter(final SAMReadGroupRecord readGroup) {
```
again, what's the need for added methods used only once elsewhere?

geoffjentry replied:
Some would say that methods ideally should only be a few lines long and anything longer should be broken down. I have my beers with the guy but IMO this is one that uncle bob got right

geoffjentry replied:
And by beers I mean beefs. Clearly one can see which word I type more frequently on my iphone

nh13 replied:
I think in this case a simple lookup as a private method is too much.

geoffjentry replied:
Now that I'm looking at code and not just commenting on comments I agree with you here.

nh13 commented on src/java/picard/sam/revertsam/PerReadGroupSorter.java:
```diff
@@ -0,0 +1,52 @@
+        this.iterators = createIterators(sorterMap.values());
+    }
+
+    private static List<PeekableIterator<SAMRecord>> createIterators(final Collection<SortingCollection<SAMRecord>> sorters) {
```
put this in the constructor?

nh13 commented on src/java/picard/sam/revertsam/RevertSam.java:
```diff
@@ -148,75 +144,51 @@ public static void main(final String[] args) {
         // Build a sorting collection to use if we are sanitizing
         ////////////////////////////////////////////////////////////////////////////
-        final SortingCollection<SAMRecord> sorter;
+        RevertSamSorter sorter = null;
```
Instead of a whole new set of classes, why not just two maps:
```
final Map<String, PeekableIterator<SAMRecord> iterators;
final Map<String, SortingCollection<SAMRecord> sorters;
```
In the case of one output, just have them map to the same iterator or sorter (you'll need to unique them later), or just condition on `OUTPUT_BY_READGROUP` laster.  There are probably other ways, but are three new classes to store two simple maps necessary?

nh13 commented on src/java/picard/sam/revertsam/RevertSam.java:
```diff
@@ -240,37 +212,97 @@ protected int doWork() {
+                rec.setNotPrimaryAlignmentFlag(false);
+                rec.setProperPairFlag(false);
+                rec.setReadUnmappedFlag(true);
+
```
newline

nh13 commented on src/java/picard/sam/revertsam/RevertSam.java:
```diff
@@ -240,37 +212,97 @@ protected int doWork() {
+            }
+
+            // Then remove any mate flags and info related to alignment
+            if (rec.getReadPairedFlag()) {
```
I would prefer that *all* mate information is removed, even if the `getReadPairedFlag` is not set.  We are trying to sanitize/revert the SAM, perhaps from a bad or inconsistent state.

dshiga replied:
This is not new code. It just got shifted to a different spot within RevertSam. I could remove the conditional, but it would mean changing behavior that some users may have come to expect. Should I add a parameter to determine this behavior instead? Does it belong in a different PR? I'm not sure how strict this project tends to be about maintaining backwards compatibility.

nh13 replied:
Some software *incorrectly* has mate information even when the read paired flag is not set, and it would be great if `RevertSam` removed that mate information.  If the goal of this tool is to have a clean and valid BAM with no mapping information in the output, I don't see what harm removing all this information regardless of whether or not the `getReadPairedFlag` is set.  @yfarjoun would you confirm my reasoning?

yfarjoun replied:
Yes. I agree. Mate information is mapping information and should be removed
during revert regardless of how the bam got to its sad state.

On Thu, May 5, 2016 at 3:15 PM, Nils Homer <notifications@github.com> wrote:

> In src/java/picard/sam/revertsam/RevertSam.java
> <https://github.com/broadinstitute/picard/pull/531#discussion_r62238942>:
>
> > +            // Remove all alignment based information about the read itself
> > +            rec.setReferenceIndex(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
> > +            rec.setAlignmentStart(SAMRecord.NO_ALIGNMENT_START);
> > +            rec.setCigarString(SAMRecord.NO_ALIGNMENT_CIGAR);
> > +            rec.setMappingQuality(SAMRecord.NO_MAPPING_QUALITY);
> > +
> > +            if (!rec.getReadUnmappedFlag()) {
> > +                rec.setInferredInsertSize(0);
> > +                rec.setNotPrimaryAlignmentFlag(false);
> > +                rec.setProperPairFlag(false);
> > +                rec.setReadUnmappedFlag(true);
> > +
> > +            }
> > +
> > +            // Then remove any mate flags and info related to alignment
> > +            if (rec.getReadPairedFlag()) {
>
> Some software *incorrectly* has mate information even when the read
> paired flag is not set, and it would be great if RevertSam removed that
> mate information. If the goal of this tool is to have a clean and valid BAM
> with no mapping information in the output, I don't see what harm removing
> all this information regardless of whether or not the getReadPairedFlag
> is set. @yfarjoun <https://github.com/yfarjoun> would you confirm my
> reasoning?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/531/files/99dbe60dd4f272e090c94f29ac77624d70dc50bc#r62238942>
>

nh13 commented on src/java/picard/sam/revertsam/RevertSam.java:
```diff
@@ -240,37 +212,97 @@ protected int doWork() {
+        }
+    }
 
+    private long[] sanitize(
```
You don't use the `RevertSamSorter` here, so I am more convinced `RevertSamSorter` can be simplified.

nh13 commented on src/java/picard/sam/revertsam/RevertSam.java:
```diff
@@ -359,56 +379,19 @@ public boolean filterOut(final SAMRecord first, final SAMRecord second) {
-                rec.setOriginalBaseQualities(null);
-            }
-        }
+    private String getDefaultExtension(final String input) {
```
`/dev/stdin`? Urls?

dshiga replied:
I don't understand the question. Are you asking for getDefaultExtension to be able to handle input from stdin and urls? I have inlined this method now so maybe this is a non issue.

nh13 replied:
If I want to output by read group, but the input comes from either a URL or `/dev/stdin`, then calling `getDefaultExtension` on `INPUT` will return `.bam`.  Also, this doesn't support `.cram`.  I would add documentation that says the output is BAM by default unless the input ends with `.sam` or .cram`.

nh13 replied:
Just to me more explicit, I think defaulting to output `.bam` is reasonable but `.cram` isn't supported and some docs are needed.

dshiga replied:
Okay - add documentation in the doc string for the OUTPUT param, or in the usage details documentation? Or did you mean a code comment where we define this default?

nh13 replied:
I think in the usage, since you will want two usage examples anyhow, one for the outputting to one file, and one when outputting per-read-group. @vdauwera does having two examples make sense?

vdauwera replied:
I think in this case it makes sense, yes. In general we should be wary of including multiple examples just to show how a particular argument is invoked (eg how we went overboard with SelectVariants in GATK) but these are likely to be equally important use cases. So +1 from me.

nh13 commented on src/java/picard/sam/revertsam/RevertSam.java:
```diff
@@ -359,56 +379,19 @@ public boolean filterOut(final SAMRecord first, final SAMRecord second) {
 
-        if (REMOVE_DUPLICATE_INFORMATION) {
-            rec.setDuplicateReadFlag(false);
+    private void overrideSample(final SAMFileHeader header, final String sampleAlias) {
```
`overwriteSample`

nh13 commented on src/java/picard/sam/revertsam/RevertSam.java:
```diff
@@ -359,56 +379,19 @@ public boolean filterOut(final SAMRecord first, final SAMRecord second) {
-            for (final String tag : ATTRIBUTE_TO_CLEAR) {
-                rec.setAttribute(tag, null);
-            }
+    private void overrideLibrary(final SAMFileHeader header, final String libraryName) {
```
`overwriteLibrary`?

nh13 commented on src/java/picard/sam/revertsam/RevertSamSorter.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Allows callers to call the sorter in the same way, regardless of whether it's a SingleSorter or a PerReadGroupSorter.
+ */
+public interface RevertSamSorter {
```
make package protected if this stays around.

nh13 commented on src/java/picard/sam/revertsam/RevertSamUtil.java:
```diff
@@ -0,0 +1,188 @@
+/**
+ * Functions for creating writers, headers, and other objects needed by RevertSam.
+ */
+public class RevertSamUtil {
```
what's the benefit of shoving these methods into a generic `Util` class that is so tied to `RevertSam` that it's in the name?

nh13 commented on src/java/picard/sam/revertsam/RevertSamUtil.java:
```diff
@@ -0,0 +1,188 @@
+        return readGroupList;
+    }
+
+    private static SAMFileHeader createOutHeaderForReadGroup(
```
I am just not convinced that this makes it more readable to have a private static method that is used only once.

nh13 commented on src/java/picard/sam/revertsam/RevertSamUtil.java:
```diff
@@ -0,0 +1,188 @@
+
+    private final static Log log = Log.getInstance(RevertSamUtil.class);
+
+    public static List<ReadGroup> makeReadGroupList(
```
If this was a public class and method, it should be documented.

nh13 commented on src/java/picard/sam/revertsam/RevertSamUtil.java:
```diff
@@ -0,0 +1,188 @@
+        return outHeader;
+    }
+
+    public static RevertSamWriter makeSingleOrPerReadGroupWriter(
```
This seems like a lot of extra code and effort to hide if we are writing to one output file or one per read group.  What does this buy you?  It's not like `RevertSam` would be too much more complex if you had `Map<String, SAMFileWriter> writers`.

nh13 commented on src/java/picard/sam/revertsam/RevertSamUtil.java:
```diff
@@ -0,0 +1,188 @@
+        return writerMap;
+    }
+
+    public static SAMFileHeader createSingleOutHeader(
```
It was confusing that this is called `createSingleOutHeader` yet its output is eventually used only by `makeSingleOrPerReadGroupWriter` and only when creating `innerWriter`.  Seems hard to follow.  Why not just put it in `createWriterMap`?

dshiga replied:
Having simplified by removing classes, etc., hopefully this is clearer now. singleOutHeader is used for creating the writer and the sorter when OUTPUT_BY_READGROUP=false.

nh13 commented on src/java/picard/sam/revertsam/RevertSamUtil.java:
```diff
@@ -0,0 +1,188 @@
+        return outHeader;
+    }
+
+    public static Map<SAMReadGroupRecord, FastqQualityFormat> createReadGroupFormatMap(
```
Again, I question why this is in a Util class.

nh13 commented on src/java/picard/sam/revertsam/RevertSamUtil.java:
```diff
@@ -0,0 +1,188 @@
+        return readGroupToFormat;
+    }
+
+    public static OutputMap createOutputMap(
```
can be removed after getting rid of the `OutputMap` class

nh13 commented on src/java/picard/sam/revertsam/RevertSamUtil.java:
```diff
@@ -0,0 +1,188 @@
+    private static Map<String, SortingCollection<SAMRecord>> createSorterMap(
+            final List<ReadGroup> readGroups,
+            final int maxRecordsInRam) {
+        
```
newline

nh13 commented on src/java/picard/sam/revertsam/SingleSamFileWriter.java:
```diff
@@ -0,0 +1,27 @@
+    public void close() {
+        writer.close();
+    }
+
```
nl

nh13 commented on src/java/picard/sam/revertsam/SingleSamFileWriter.java:
```diff
@@ -0,0 +1,27 @@
+package picard.sam.revertsam;
```
licenses again

nh13 commented on src/java/picard/sam/revertsam/ValidationUtil.java:
```diff
@@ -0,0 +1,124 @@
+/**
+ * Helper functions for validating RevertSam parameters. 
+ */
+public class ValidationUtil {
```
inner class in `RevertSam`?  Just have the methods in `RevertSam`?  package protected class?

nh13 commented on src/java/picard/util/TabbedTextFileWithHeaderParser.java:
```diff
@@ -145,6 +149,8 @@ public boolean hasColumn(final String columnLabel) {
         return columnLabelIndices.keySet();
     }
+    
+    public List<String> columnLabelsList() { return Collections.unmodifiableList(columnLabels); }
```
no docs, and can you make a list from `columnLabelIndices.keySet()` so you don't need a new member?

nh13 commented:
@dshiga @yfarjoun this is certainly a lot of effort, and I definitely see why you would want this tool, since regularly one receives (mapped) BAM files that need to be reverted or sanitized to create "unmapped" BAMs as input to analysis pipeline.  Nonetheless, I think the code could be simplified with some more thought.  I suggested a way that may not be the best way but is a start to the discussion.

nh13 commented:
Also, I omitted marking all missing licenses, the many places where classes or methods scope should be changed and documentation added, as well the many `private static` methods that are used only once.

yfarjoun commented:
Thank-you very much @nh13 for the through review and suggestions!

On Wed, May 4, 2016 at 1:49 AM, Nils Homer <notifications@github.com> wrote:

> Also, I omitted marking all missing licenses, the many places where
> classes or methods scope should be changed and documentation added, as well
> the many private static methods that are used only once.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/531#issuecomment-216750786>
>

geoffjentry commented:
I generally agree with @nh13 that this could be simplified down a fair bit. Nils biased this witness a bit too much so I found myself agreeing w/ him (egads!) vs. being able to come up with another alternative. Nils is smarter than me anyways.

dshiga commented:
Thanks @nh13, lots of good food for thought here.

I have moved things in the direction you suggested. There are no longer any new classes outside of RevertSam. Many of the methods that were in separate classes before are now just private methods of RevertSam. I have inlined all the methods you suggested inlining.

I had broken things out into independent classes and methods because I personally find that approach easier to read and understand. But based on feedback from you and others, I'm clearly in the minority in seeing it that way, at least as far as this PR is concerned. Since it seemed to be making it harder to read for most people, which was the exact opposite of my intent, I've reversed course.

My guess is that it was particularly unwelcome to have to jump around to a lot of different files to see all the code. So all the new code is in RevertSam.

I did find that some of the new methods naturally go together, and have grouped them into three new inner classes: RevertSamWriter, RevertSamSorter, and ValidationUtil. My hope is that grouping methods together in this way will make it easier for developers to navigate around in RevertSam and to safely ignore code that's not relevant to a particular task. If it hinders clarity at all, that would be good to know.

The refactoring has also had these happy effects:

- No need for adding licenses now that there are no new non-test files
- No new public classes or methods, so no need to add docs for them

There are a few other comments I will address individually.

dshiga commented:
Okay, I just pushed some more changes, which should address the last two outstanding comments.

- Added documentation about the output format in the usage details string
- Removed the conditional so that we will remove mate information for all reads regardless of whether its read paired flag is set

Does anyone have any further feedback, or can I get a thumbs up?

nh13 commented:
@dshiga do you mind if I do one more review as likely your changes were significant?  If you could give me this weekend I can get to it by EOD Monday.

dshiga commented:
@nh13 that would be great, thanks!

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -78,18 +82,43 @@
             "restoring the original quality scores of a BAM file that has already undergone base quality score recalibration (BQSR) if the" +
             "original qualities were retained." +
-            "<h4>Usage example:</h4>" +
+            "<h4>Usage example 1:</h4>" +
```
Perhaps instead of `...example 1` state `example single output`.

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -78,18 +82,43 @@
             "     O=reverted.bam" +
             "</pre>" +
+            "Output format is BAM by default, or SAM if the input path ends with '.sam'." +
+            "<h4>Usage example 2:</h4>" +
```
Perhaps instead of `...example 2` state `example output per read group`.

vdauwera replied:
I agree, specifying in the title of the example what distinguishes/characterizes it is very helpful.

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -148,76 +177,57 @@ public static void main(final String[] args) {
+        final Map<String, File> outputMap;
+        final Map<String, SAMFileHeader> headerMap;
+        if (OUTPUT_BY_READGROUP) {
+            String defaultExtension = INPUT.toString().endsWith(".sam") ? ".sam" : ".bam";
```
add support for writing a cram with `.cram`.

nh13 replied:
`final`

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -78,18 +82,43 @@
+            "     OUTPUT_BY_READGROUP=true \\<br />" +
+            "     OUTPUT_MAP=reverted_bam_paths.tsv" +
+            "</pre>" +
+            "Will output a BAM/SAM file per read group. By default, all outputs will be in BAM format. " +
```
document `.cram` support.

nh13 commented:
@ktibbett I would also like a 👍 that this PR couldn't be solved by running `SplitSamByLibrary` then the previous version of `RevertSam` on each output BAM (which could be parallelized).

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -148,76 +177,57 @@ public static void main(final String[] args) {
-        if (SANITIZE && SORT_ORDER != SortOrder.queryname) {
-            return new String[]{"SORT_ORDER must be queryname when sanitization is enabled with SANITIZE=true."};
-        }
+        List<String> errors = new ArrayList<String>();
```
final?

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -148,76 +177,57 @@ public static void main(final String[] args) {
+        if (SAMPLE_ALIAS != null) overwriteSample(inHeader.getReadGroups(), SAMPLE_ALIAS);
+        if (LIBRARY_NAME != null) overwriteLibrary(inHeader.getReadGroups(), LIBRARY_NAME);
+        final SAMFileHeader singleOutHeader = createOutHeader(inHeader, SORT_ORDER, REMOVE_ALIGNMENT_INFORMATION);
+        for (final SAMReadGroupRecord readGroup : inHeader.getReadGroups()) {
```
covert to a stream?

dshiga replied:
I'll change to use forEach here - don't think there's any need to use a stream.

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -148,76 +177,57 @@ public static void main(final String[] args) {
+            ValidationUtil.assertAllReadGroupsMapped(outputMap, inHeader.getReadGroups());
+            headerMap = createHeaderMap(inHeader, SORT_ORDER, REMOVE_ALIGNMENT_INFORMATION);
+        } else {
+            outputMap = null;
```
Do we want nulls anymore or should we use `Optional`?

dshiga replied:
In general I like getting away from nulls, however I don't think Optional is helpful in this case. We should only ever try to access outputMap and headerMap if OUTPUT_BY_READGROUP is true. If we ever try to access them when OUTPUT_BY_READGROUP is false then there is a bug in the code somewhere, and it's appropriate to blow up in that case. I'd like to avoid masking the bug with a more obscure stack trace related to Optional, or silently succeeding but producing garbage.

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -240,37 +250,96 @@ protected int doWork() {
-            long total = 0, discarded = 0;
-            final PeekableIterator<SAMRecord> iterator = new PeekableIterator<SAMRecord>(sorter.iterator());
-            final Map<SAMReadGroupRecord, FastqQualityFormat> readGroupToFormat = new HashMap<SAMReadGroupRecord, FastqQualityFormat>();
+            long[] sanitizeResults = sanitize(readGroupToFormat, sorter, out);
```
`final`.  IntelliJ should highlight these as yellow if you are using it.

geoffjentry replied:
I don't know if this is the case for @dshiga but I learned recently that for people using the dark theme it's kinda hard to spot those warnings, the yellow shade they use there is subtle.

nh13 replied:
I use the dark theme.

nh13 replied:
Also, the top of the source will show yellow too.

dshiga replied:
I had those "could be final" warnings turned off. No longer.

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -240,37 +250,96 @@ protected int doWork() {
-                };
-                readGroupToFormat.put(rg, QualityEncodingDetector.detect(QualityEncodingDetector.DEFAULT_MAX_RECORDS_TO_ITERATE, new FilteringIterator(reader.iterator(), filter), RESTORE_ORIGINAL_QUALITIES));
-                CloserUtil.close(reader);
+            if (discarded / (double) total > MAX_DISCARD_FRACTION) {
```
use `discardRate`?

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -240,37 +250,96 @@ protected int doWork() {
+            rec.setCigarString(SAMRecord.NO_ALIGNMENT_CIGAR);
+            rec.setMappingQuality(SAMRecord.NO_MAPPING_QUALITY);
+
+            if (!rec.getReadUnmappedFlag()) {
```
Just like clearing all the mate information, I think that these should all be set regardless of the `getReadUnmappedFlag` flag.

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -240,37 +250,96 @@ protected int doWork() {
+            rec.setMateUnmappedFlag(true);
 
+            // And then remove any tags that are calculated from the alignment
+            for (final String tag : ATTRIBUTE_TO_CLEAR) {
```
`forEach`?

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -240,37 +250,96 @@ protected int doWork() {
+    }
+
+    private long[] sanitize(
+            final Map<SAMReadGroupRecord, FastqQualityFormat> readGroupToFormat,
```
formatting

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -359,56 +416,344 @@ public boolean filterOut(final SAMRecord first, final SAMRecord second) {
     }
 
+    private void overwriteSample(final List<SAMReadGroupRecord> readGroups, final String sampleAlias) {
+        for (final SAMReadGroupRecord rg : readGroups) {
```
`forEach`?

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -359,56 +416,344 @@ public boolean filterOut(final SAMRecord first, final SAMRecord second) {
+    }
+
+    private void overwriteLibrary(final List<SAMReadGroupRecord> readGroups, final String libraryName) {
+        for (final SAMReadGroupRecord rg : readGroups) {
```
`forEach`?

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -359,56 +416,344 @@ public boolean filterOut(final SAMRecord first, final SAMRecord second) {
+        }
+    }
+
+    Map<String, File> createOutputMap(
```
could this be `static`?

nh13 commented on src/java/picard/sam/RevertSam.java:
```diff
@@ -359,56 +416,344 @@ public boolean filterOut(final SAMRecord first, final SAMRecord second) {
+            final File outputPath = new File(output);
+            outputMap.put(id, outputPath);
+        }
+        parser.close();
```
`CloserUtil.close`

nh13 commented:
@dshiga this looks a lot better, could you fix up some of the formatting issues and then 👍 .

ktibbett commented:
@nh13 , this was at my request, so I do 👍 .  This output-by-readgroup functionality is analogous to what we do elsewhere, and (though I do take your point about SplitSamByLibrary, whose existence I had forgotten) this implementation means we read the file once rather than twice.

nh13 commented:
I'd like to point out that due to sorting you may read the contents of the file more than once, but you are correct, the original file is only read once.  👍

dshiga commented:
@nh13 I've addressed all your comments now. I will merge at the end of the day unless you shout.

yfarjoun commented:
I think that protocol should be that we wait for an explicit response
rather than assume silent acquiescence. If @nh13 takes a very long time to
respond, we can discuss how to move forward...but I doubt that would happen.

On Mon, May 9, 2016 at 3:37 PM, David Shiga <notifications@github.com>
wrote:

> @nh13 <https://github.com/nh13> I've addressed all your comments now. I
> will merge at the end of the day unless you shout.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/531#issuecomment-217966410>
>

dshiga commented:
Okay, I'll wait -- I'm not trying to sneak this in or anything! His last comment asked for some formatting changes, then gave the 👍. In my experience that means go ahead and merge after you've made those changes. Is the convention different here?

nh13 commented:
@dshiga 
![zgpk3](https://cloud.githubusercontent.com/assets/846274/15126823/ef1e111e-1601-11e6-8676-1cbfdeb76b8c.gif)

dshiga commented:
Thanks @nh13, merging away!

yfarjoun commented:
@dshiga You are correct, I was slightly oversensitive here, and you would have been 💯 to merge your formatting changes. This PR was big and chatty that I wanted to make sure everything in on the up-and-up!

Thanks to all who helped here!!
